UI/CUIControls.cpp

```diff
@@ -1924,3 +1928,139 @@ void RotatingGraphic::Render() {
     glPopClientAttrib();
     glPopMatrix();
 }
+
+////////////////////////////////////////////////
+// CUIToggleButton
+////////////////////////////////////////////////
+CUIToggleRepresenter::CUIToggleRepresenter(const std::string& tooltip,
+                                           boost::shared_ptr<GG::SubTexture> icon,
+                                           const GG::Clr& highlight_clr) :
+                                           m_unchecked_icon(icon),
+                                           m_checked_icon(icon),
+                                           m_unchecked_color(highlight_clr),
+                                           m_checked_color(highlight_clr),
+                                           m_tooltip(tooltip)
+{}
+
+boost::shared_ptr<GG::SubTexture> CUIToggleRepresenter::GetUncheckedIcon() const
+{ return m_unchecked_icon; }
+
+boost::shared_ptr<GG::SubTexture> CUIToggleRepresenter::GetCheckedIcon() const
+{ return m_checked_icon; }
+
+const bool CUIToggleRepresenter::HasSharedIcon() const
+{ return (m_checked_icon == m_unchecked_icon); }
+
+const GG::Clr& CUIToggleRepresenter::GetUncheckedColor() const
+{ return m_unchecked_color; }
+
+const GG::Clr& CUIToggleRepresenter::GetCheckedColor() const
+{ return m_checked_color; }
+
+void CUIToggleRepresenter::SetTooltip(const std::string& tooltip)
+{ m_tooltip = tooltip; }
+
+/** \brief Set the SubTexture for both checked and unchecked states
+ * 
+ * @param[in] icon SubTexture to assign
+ */
+void CUIToggleRepresenter::SetIcon(boost::shared_ptr<GG::SubTexture> icon) {
+    m_checked_icon = icon;
+    m_unchecked_icon = icon;
+}
+
+/** Set the SubTexture for the Unchecked state
+ * 
+ * @param[in] icon  SubTexture to assign
+ */
+void CUIToggleRepresenter::SetUncheckedIcon(boost::shared_ptr<GG::SubTexture> icon)
+{ m_unchecked_icon = icon; }
+
+/** \brief Set the SubTexture for the Checked state
+ * 
+ * @param[in] icon SubTexture to assign
+ */
+void CUIToggleRepresenter::SetCheckedIcon(boost::shared_ptr<GG::SubTexture> icon)
+{ m_checked_icon = icon; }
+
+void CUIToggleRepresenter::SetColor(GG::Clr c) {
+    m_unchecked_color = c;
+    m_checked_color = c;
+}
+
+void CUIToggleRepresenter::SetUncheckedColor(GG::Clr c)
+{ m_unchecked_color = c; }
+
+void CUIToggleRepresenter::SetCheckedColor(GG::Clr c)
+{ m_checked_color = c; }
+
+const bool CUIToggleRepresenter::HasSharedColor() const
+{ return (m_unchecked_color == m_checked_color); }
+
+void CUIToggleRepresenter::Render(const CUIToggleButton& button) const {
+    bool render_checked(!HasSharedIcon() && (!button.Checked()^button.IsMouseover()));  // use specific icon if not shared, swap on mouseover
```

Comment:
Why do you need this testing of whether there is a shared icon? Why not just render whichever texture is appropriate for the state of the button (checked or unchecked)?